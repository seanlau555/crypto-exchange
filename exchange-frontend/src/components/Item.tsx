import { DivEvent } from '../types'
import { useGetLatestPrice } from '../services'
import { Box } from '@chakra-ui/react'

function Item({
  current,
  to,
  selectedTicker,
  onClick,
}: {
  current: string
  to: string
  selectedTicker: string
  onClick: (evt: DivEvent) => void
}) {
  const { data } = useGetLatestPrice(current, to)
  const price = `Price: ${to} ${data}`
  const exchange = current + to

  return (
    <Box
      display="block"
      cursor="pointer"
      padding="4px 16px"
      borderWidth="2px"
      borderColor={selectedTicker === exchange ? 'teal.100' : 'white'}
      fontWeight={selectedTicker === exchange ? '600' : '500'}
      background={selectedTicker === exchange ? 'teal.50' : 'white'}
      color={selectedTicker === exchange ? 'teal.700' : 'gray.700'}
      borderRadius="0px"
      onClick={onClick}
      id={exchange}
      data-testid={exchange}
      fontSize="sm"
    >
      {exchange}, {price}
    </Box>
  )
}
export default Item
