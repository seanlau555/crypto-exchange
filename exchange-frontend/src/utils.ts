import axios, { AxiosResponse } from 'axios'
import { Bar } from './types'

const Utils = {
  async getAuthToken(oauth_code: string) {
    // returns Authorization Token once we have our OAuth token
    const body = {
      grant_type: 'authorization_code',
      code: oauth_code,
      client_id: import.meta.env.VITE_REACT_APP_CLIENT_ID,
      client_secret: import.meta.env.VITE_REACT_APP_CLIENT_SECRET,
      redirect_uri: import.meta.env.VITE_REACT_APP_REDIRECT_URI,
    }
    // encode data into form encoding
    const encodedBody = Object.keys(body)
      .map((key: string) => `${key}=${encodeURIComponent(body[key])}`)
      .join('&')
    console.log(body)
    // submit POST request
    const response = await axios({
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      url: 'https://api.alpaca.markets/oauth/token',
      data: encodedBody,
    })

    const { data } = response
    window.localStorage.setItem('auth-token', data)
    return data.access_token
  },

  parseResponse(response: AxiosResponse): Array<Bar> | [] {
    const { bars } = response.data
    if (bars) {
      const data = []
      for (let i = 0; i < bars.length; i++) {
        const bar = bars[i]
        const point = {
          date: bar.t,
          open: bar.o,
          low: bar.l,
          high: bar.h,
          close: bar.c,
          volume: bar.v,
        }
        data.push(point)
      }
      return data
    }
    return []
  },
}
export default Utils
