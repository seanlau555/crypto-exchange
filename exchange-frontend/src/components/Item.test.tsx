import { describe, expect, it } from 'vitest'
import Item from './Item'
import { createWrapper } from '../test/utils'

describe('<Item />', () => {
  it('render Item', () => {
    const { container } = createWrapper(
      <Item
        current={'BTC'}
        to={'USD'}
        selectedTicker={'BTCUSD'}
        onClick={() => {}}
      />,
    )
    expect(container).toMatchSnapshot()
  })
})
