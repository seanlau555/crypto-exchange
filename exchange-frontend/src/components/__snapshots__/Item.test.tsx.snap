// Vitest Snapshot v1

exports[`<Item /> > render Item 1`] = `
<div>
  <div
    class="css-1j9jo68"
    data-testid="BTCUSD"
    id="BTCUSD"
  >
    BTCUSD
    , 
    Price: USD undefined
  </div>
</div>
`;
